flocker/control/_protocol.py

```diff
@@ -253,7 +315,19 @@ def __init__(self, agent):
         CommandLocator.__init__(self)
         self.agent = agent
 
+    @property
+    def logger(self):
+        """
+        ``with_eliot_context`` assumes that the decorated method's bound
+        instance has a logger, which it could, but for testing, it needs to use
+        the same logger as the supplied agent.
+
+        XXX This seems ugly.
+        """
+        return self.agent.logger
```

Comment:
The correct comment to make here, then, is that the `logger` attribute isn't part of `IConvergenceAgent`.  And `FakeAgent` is an unverified fake so nothing demonstrates that it's compatible with the useful implementation of `IConvergenceAgent` (`ControlAMPService`).

That's too bad.  And more generally, `FakeAgent` is an example of not-very-good testing methodology that we should avoid.  It's basically a minimally functional, completely custom mock.  So it has the correctness drawbacks of mocks but without any of the convenience of mocks.  Sad.

The interface should be fixed and a test added that can be shared between `IConvergenceAgent` implementations demonstrating that (I suppose) they have a `logger` attribute that is an instance of `Logger`.